### ISSUE TITLE
CollectionAssert.AllItemsAreUnique() very slow

### ISSUE THREAD

**Issue by rvhuang** [nunit/nunit#2494]
Consider following simple test case:

```csharp
// using NUnit.Framework;
        [TestCase(10000)]
        public void TestMethod1(int range)
        {
            CollectionAssert.AllItemsAreUnique(Enumerable.Range(0, range).ToArray());
        }
```
it takes about 14 seconds to complete.

However, applying MS Test Framework to same test case, it takes about 12 ms only to finish.

```csharp
// using Microsoft.VisualStudio.TestTools.UnitTesting;
        [TestMethod]
        public void TestMethod1()
        {
            CollectionAssert.AllItemsAreUnique(Enumerable.Range(0, 10000).ToArray());
        }
```

The gap between two should not be that huge.

**Comment by ChrisMaddock** [comment]
NUnit uses a different way to test for uniqueness, than MSTest.

MSTest uses simple object equality. NUnit equality tests go through [NUnitEqualityComparer ](https://github.com/nunit/nunit/blob/e7c7826da8b064cfcc518b1035e1c270588abc79/src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs)- which has extra complexity.

Looking at that class however, there may be low hanging fruit. Before comparing for object equivalence, we create an empty List of FailurePoint objects, which doesn't seem to be used. We then do two null checks prior to object.ReferenceEquals(), which I think are unnecessary. 

I'll dig in to this one more later.

**Comment by rvhuang** [comment]
@ChrisMaddock Thanks for looking into the issue. We have a test case which requires to check uniqueness of 10k items and it cannot finish within one minute. The performance improvement shall help us a lot.

**Comment by CharliePoole** [comment]
@ChrisMaddock I'm not looking at the code but I believe we accept an IEnumerable rather than an ICollection. You. Might be able to create a special case for collections. Of course we are also doing a lot of boxing in the case of an int value.

**Comment by ChrisMaddock** [comment]
@CharliePoole - we do, but I'm not sure what sort of special case you're thinking of? Would love to hear, if you get a chance to look later!

**Comment by CharliePoole** [comment]
A simple thing to try would be (in `Matches`)

```C#
IList list = actual as IList ?? new List<object>(actual);
```

That way, if an IList were passed in, we would not create a new one.

Following that (if tests showed it necessary) I'd use a generic helper to eliminate boxing. I'm busy on parallelization stuff or I'd give it a try myself.

**Comment by ChrisMaddock** [comment]
This is a good candidate for [Accidentally Quadratic](https://accidentallyquadratic.tumblr.com/)! (If not purposefully quadratic...) Question is now, how we avoid that, whilst still using NUnitEqualityComparer.

Out of time for now, will come back to it. My first plan is to add some efficiencies to `NUnitEqualityComparer.AreEqual()` - this is low level so should be as efficient as possible - and there's some low hanging fruit here. After that, depending on the effect that has on runtime, I'll come back to look at UniqueItemsConstrant specifically, and see what can be done there.

**Comment by CharliePoole** [comment]
Hard to see how comparing each item to each other item could be any less than O(n*(n-1))

I'll take a quick shot at an idea I have.

**Comment by jnm2** [comment]
The typical thing people to is a hash table which is likely where MSTest gets its speed, but we would have to be very careful how we calculated the hashes for that so that no two hashes are inequal if `NUnitEqualityComparer` would consider them equal.

**Comment by jnm2** [comment]
My bad, the hash table is technically O(1).

**Comment by jnm2** [comment]
Okay, I'm reading that checking for element uniqueness with a hash table is O(n).

**Comment by jnm2** [comment]
For now I'd say don't use `CollectionAssert.AllItemsAreUnique` for this scenario. You often have to write specialized assertion code in specialized scenarios. In your case, all you need is a `new HashSet<string>(StringComparer.Whichever)` and you can find and report non-unique instances.

Or this, if you want repeat counts:
```cs
var nonuniqueItems = list
	.GroupBy(str => str, (key, group) => (value: key, repeats: group.Count()), StringComparer.Whichever)
	.Where(result => result.repeats > 1)
	.Select(result => $"{result.value}: {result.repeats} times");

Assert.That(nonuniqueItems,	Is.Empty);
```

And if lambdas aren't efficient enough, you can translate this to super tight `Dictionary<string, int>` code.

**Comment by CharliePoole** [comment]
@ChrisMaddock I added my name as a co-assignee. I'm working at the top end and I'll leavE `NUnitEqualityComparer` for you. PR shortly.

**Comment by CharliePoole** [comment]
@ChrisMaddock Are you sure MSTest uses object equality?

**Comment by CharliePoole** [comment]
Some choices we appear to have here...
1. Leave it alone - doesn't seem to be acceptable given how slow it is.
2. Document that we __do not__ use NUnit equality for this constraint.
3. Document that we __do not__ use NUnit Equality for the `CollectionAssert` call and stop using the constraint in that case.
4. Analyze actual to decide whether a more efficient approach can be used in a given situation. FOr example, if actual is `IEnumerable<T>` then we know all elements are of type `T`. If `T` is not a Type we handle specially, we can use a simpler algorithm. IOW, use the efficient algorithm only if we can guarantee it is correct.
5. Figure out how to use NUnit equality comparer with the hashing approach. Somebody other than me could give this a shot.

**Comment by ChrisMaddock** [comment]
@CharliePoole - Looks that way to me: https://github.com/Microsoft/testfx/blob/e3d6e822df12fed112a9c68311a11fac8a4cf914/src/TestFramework/MSTest.Core/Assertions/CollectionAssert.cs#L338

I'd personally considered changing functionality here too much of a breaking change. I agree we should work to improve this, however, if nothing can be done, I would consider leaving it unchanged preferable to changing our measure of 'unique'. Of course, there's more voices than just mine in this decision. 😄

Option 4 is a good shout. Option 5 is something else I considered, but sounds pretty messy! 

Another option, which I have no idea to the feasibility of, is introduce a typed NUnitEqualityComparer. The current version checks all tests of equality for all types - meaning asserting that something is _not_ equal as we are doing here is pretty expensive. The method current takes two objects - I wonder if we could introduce specific overloads to handle different types. This is similar to option (4) - but tackling the issue at a lower level.

**Comment by jnm2** [comment]
I think the same as @ChrisMaddock. My personal preference would be a combination of options 4 and 5 and Chris's idea, and those things don't all have to be tackled at the same time.

**Comment by rvhuang** [comment]
> Analyze actual to decide whether a more efficient approach can be used in a given situation. FOr example, if actual is IEnumerable<T> then we know all elements are of type T. If T is not a Type we handle specially, we can use a simpler algorithm. IOW, use the efficient algorithm only if we can guarantee it is correct.

Maybe it can do a further check which is checking if type `T` implements `IEquatable<T>`. If it does, the whole test can easily be done by utilizing `EqualityComparer<T>.Default`.

**Comment by CharliePoole** [comment]
In this PR I'm limiting myself to fixes in the constraint itself. Longer term, I agree we should look at how we do equality in general.

**Comment by CharliePoole** [comment]
The top-down fix is ready to merge in PR #2501. I'm taking my name off this issue and leaving @ChrisMaddock to decide if further work at the back end is necessary.

**Comment by ChrisMaddock** [comment]
Let's close this when the PR is merged. I think any changes in NUnitDefaultEqualityComparer will be micro-optimisations, which I'd like to do anyway, but are a separate issue.

**Comment by mikkelbu** [comment]
Following @ChrisMaddock's comment I've closed this issue since #2501 has been merged
